### ISSUE TITLE
Wrong `self` and `pkgs` names in `all-packages.nix`

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#34881]
## Issue description

The `pkgs/top-level/all-packages.nix` is essentially an overlay, but its parameters are called `self` and `pkgs`, which is wrong. `self` in that file is actually what is normally called `super` and `pkgs` is what is normally called `self`.

This causes confusion.

I suppose this itself is a legacy problem and changing these names will break many nixpkgs forks. Fixing this straight away will produce problems that are not easy to relate to this change, so I suggest we follow a simple migration plan

### Steps to reproduce

Try to use `self.runCommand` in `all-packages.nix`.

or

Add a line `xself = self;` and run

    $ nix-instantiate --eval --expr '(import ./. {}).xself.runCommand'
    error: attribute ‘runCommand’ missing
    $ nix-instantiate --eval --expr '(import ./. { overlays = [ (self: super: { foo = true; }) ]; }).xself.runCommand'
    error: attribute ‘foo’ missing

## Migration plan

  1. Rename to `pkgsSelf` and `pkgsSuper`; create aliases for `self` and `nixpkgs` that emit warnings using `builtins.trace`. 
  2. Remove the aliases.
  3. (optional) Alias `pkgsSelf` and `pkgsSuper` to `self` and `super`.
  4. (optional) Remove `pkgsSelf` and `pkgsSuper`.

Code:

  0. `self: pkgs:`
  1. `pkgsSuper: pkgsSelf: let pkgs = deprecated pkgsSelf; self = deprecated pkgsSuper; in` (yes, the parameters are in flipped order)
  2. `super: self: let pkgsSuper = deprecated super; pkgsSelf = deprecated self; in`
  3. `super: self:`

where `deprecated` is a meta-expression that represents the identity function with the side effect of emitting the right warning using `builtins.trace`. (I've omitted the actual warnings for conciseness)

Of course, the contents of `all-packages.nix` will be updated to reflect the current best practice with each of these changes.

We might also want to fix the mysterious flipped parameter order. (Which may have been caused by an implementation detail of the overlay fold that was used?) This should be easier to fix, because it would only break forks that introduce new hardcoded overlays (i.e. not via the 'public' `overlays` parameter in `/default.nix` etc)

Any opinions?

**Comment by roberth** [comment]
@fpletz, I think it'd be great if we can get step 1 (deprecation) into 18.03. What do you think?

**Comment by roberth** [comment]
`pkgs` to `self` rename is not that important. #51278 is the start of a much simplified version of this.

**Comment by roberth** [comment]
Interestingly, the first parameter is neither *self* nor *super* but the *super* of the subsequent stage...

**Comment by roberth** [comment]
@nbp, do you happen to know why [`res`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/stage.nix#L92) is needed?

**Comment by nbp** [comment]
Honestly, I do not know why it is needed, but I recall that changing it to be what overlays are expecting today, at the time of this reorganization, was a hash-changing operation.

`res` first got introduced as a `rec`, then as a special attribute in `all-packages.nix`, The fact that we did not got the same hash after going through the fix-point of a worrying issue, which is likely related to the way `stdenv` used to be defined. I do not know if the recent re-font of the `stdenv` to offer better support for cross-compilation will change or not this story.

Ideally `res` should be replaced by either `self`, `super` or `rec`, depending whether this is a package, a function or a function defined in `all-packages.nix`.

**Comment by Ericson2314** [comment]
@nbp it might have had something to do with aliases/security updates?

Anyways as I said in https://github.com/NixOS/nixpkgs/pull/51401#issuecomment-443562708 I think we should be able to do everything but the `pkgs` -> `self` rename at once without issues.

**Comment by nbp** [comment]
@nbp Yes, this certainly created aliases that I wanted to remove, but the first issue was the hash-changing aspect.

**Comment by roberth** [comment]
The bad `self` has been gone for a long time now and could be used instead of `pkgs`. However, we could consider `final` conforming with the enforced naming convention in flake `overlay` attrs.
